h2o-core/src/main/java/water/rapids/ast/prims/search/AstMatch.java

```diff
@@ -65,64 +100,124 @@ public ValFrame apply(Env env, Env.StackHelp stk, AstRoot asts[]) {
   }
 
   private static class StrMatchTask extends MRTask<CatMatchTask> {
-    String[] _values;
+    String[] _sortedValues;
     double _noMatch;
-    StrMatchTask(String[] values, double noMatch) {
-      _values = values;
+    int _startIndex;
+    IcedHashMap<String, Integer> _mapping;
+    IcedHashMap<String, Integer> _matchesIndexes;
+
+    StrMatchTask(String[] values, double noMatch, int indexes) {
+      _mapping = new IcedHashMap<>();
+      for (int i = 0; i < values.length; i++) {
+        _mapping.put(values[i], i);
+      }
+      _sortedValues = values.clone();
+      Arrays.sort(_sortedValues);
       _noMatch = noMatch;
+      _startIndex = indexes;
+      _matchesIndexes = new IcedHashMap<>();
     }
+    
     @Override
     public void map(Chunk c, NewChunk nc) {
       BufferedString bs = new BufferedString();
       int rows = c._len;
       for (int r = 0; r < rows; r++) {
-        double x = c.isNA(r) ? _noMatch : in(_values, c.atStr(bs, r).toString(), _noMatch);
+        double x = c.isNA(r) ? _noMatch : in(_matchesIndexes, _sortedValues, _mapping, c.atStr(bs, r).toString(), _noMatch, _startIndex);
         nc.addNum(x);
       }
     }
   }
 
   private static class CatMatchTask extends MRTask<CatMatchTask> {
-    String[] _values;
+    String[] _sortedValues;
+    int[] _firstMatchRow;
     double _noMatch;
-    CatMatchTask(String[] values, double noMatch) {
-      _values = values;
+    int _startIndex;
+    IcedHashMap<String, Integer> _mapping;
+    IcedHashMap<String, Integer> _matchesIndexes;
+    
+ 
+    CatMatchTask(String[] values, double noMatch, int startIndex) {
+      _mapping = new IcedHashMap<>();
+      for (int i = 0; i < values.length; i++) {
+        _mapping.put(values[i], i);
```

Comment:
I would not override the mapping entry if it is already there: we want to return only the first index

Reply:
we need a good test to capture that use-case

Reply:
Hmm, I need help understanding. The mapping is initialized only at the beginning of the MR task and used during the search.

Reply:
@maurever the thing is that nothing enforces the `values` array to be made of unique values (or I missed it).
Ffor example if you refer to the R "illuminating example"
```R
## Illuminating example about NA matching
r <- c(1, NA, NaN)
zN <- c(complex(real = NA , imaginary =  r ), complex(real =  r , imaginary = NA ),
        complex(real =  r , imaginary = NaN), complex(real = NaN, imaginary =  r ))
zM <- cbind(Re=Re(zN), Im=Im(zN), match = match(zN, zN))
rownames(zM) <- format(zN)
zM ##--> many "NA's" (= 1) and the four non-NA's (3 different ones, at 7,9,10)

#--->
          Re  Im match
      NA  NA   1     1
      NA  NA  NA     1
      NA  NA NaN     1
      NA   1  NA     1
      NA  NA  NA     1
      NA NaN  NA     1
  1+NaNi   1 NaN     7
      NA  NA NaN     1
NaN+NaNi NaN NaN     9
NaN+  1i NaN   1    10
      NA NaN  NA     1
NaN+NaNi NaN NaN     9

```
we use `match(zN, zN)` and all the `NA`s match to the first occurence, not the later ones also present in `zN`.

In your implementation `values` is an array made of all `zN` values, so there are multiple `NA` there, and therefore when we loop through the values when we build the mapping, we're only interested in the first index of each occurence, typically:
```java
for (int i = 0; i < values.length; i++) {
        if (!mapping.contains(values[i])) _mapping.put(values[i], i);
}
```
do I make sense?

Reply:
so by "we need a good test to capture that use-case", I mean we need a test where the values are not all unique

Reply:
Perfect point @sebhrusen! Thanks for the explanation.